tests/explainable/test_handlers_explanation.py

```diff
@@ -314,3 +316,236 @@ def test_SplitSubsets_two_layers():
+            .log_factor[2, 0, 0, 0, :]
+            .sum()
+            .exp()
+            == 1.0
+        )
+
+    assert torch.all(
+        trace_reverse.trace.nodes["__cause____consequent_X"]["fn"]
+        .log_factor.sum()
+        .exp()
+        == 0
+    )
+
+
+# X -> Z, Y -> Z
+def model_three_converge():
+    X = pyro.sample("X", dist.Bernoulli(0.5))
+    Y = pyro.sample("Y", dist.Bernoulli(0.5))
+    Z = pyro.sample("Z", dist.Bernoulli(torch.min(X, Y)))
+    return {"X": X, "Y": Y, "Z": Z}
+
+
+# X -> Y, X -> Z
+def model_three_diverge():
+    X = pyro.sample("X", dist.Bernoulli(0.5))
+    Y = pyro.sample("Y", dist.Bernoulli(X))
+    Z = pyro.sample("Z", dist.Bernoulli(X))
+    return {"X": X, "Y": Y, "Z": Z}
+
+
+# X -> Y -> Z
+def model_three_chain():
+    X = pyro.sample("X", dist.Bernoulli(0.5))
+    Y = pyro.sample("Y", dist.Bernoulli(X))
+    Z = pyro.sample("Z", dist.Bernoulli(Y))
+    return {"X": X, "Y": Y, "Z": Z}
+
+
+# X -> Y, X -> Z, Y -> Z
+def model_three_complete():
+    X = pyro.sample("X", dist.Bernoulli(0.5))
+    Y = pyro.sample("Y", dist.Bernoulli(X))
+    Z = pyro.sample("Z", dist.Bernoulli(torch.max(X, Y)))
+    return {"X": X, "Y": Y, "Z": Z}
+
+
+# X -> Y    Z
+def model_three_isolate():
+    X = pyro.sample("X", dist.Bernoulli(0.5))
+    Y = pyro.sample("Y", dist.Bernoulli(X))
+    Z = pyro.sample("Z", dist.Bernoulli(0.5))
+    return {"X": X, "Y": Y, "Z": Z}
+
+
+# X     Y    Z
+def model_three_independent():
+    X = pyro.sample("X", dist.Bernoulli(0.5))
+    Y = pyro.sample("Y", dist.Bernoulli(0.5))
+    Z = pyro.sample("Z", dist.Bernoulli(0.5))
+    return {"X": X, "Y": Y, "Z": Z}
+
+
+@pytest.mark.parametrize("ante_cons", [("X", "Y", "Z"), ("X", "Z", "Y")])
+@pytest.mark.parametrize(
+    "model",
+    [
+        model_three_converge,
+        model_three_diverge,
+        model_three_chain,
+        model_three_complete,
+        model_three_isolate,
+        model_three_independent,
+    ],
+)
+def test_eq_neq_three_variables(model, ante_cons):
+    ante1, ante2, cons = ante_cons
+    with ExtractSupports() as supports:
+        model()
+
+    with MultiWorldCounterfactual() as mwc:
+        with SearchForExplanation(
+            supports=supports.supports,
+            antecedents={ante1: torch.tensor(1.0), ante2: torch.tensor(1.0)},
+            consequents={cons: torch.tensor(1.0)},
+            witnesses={},
+            alternatives={ante1: torch.tensor(0.0), ante2: torch.tensor(0.0)},
+            antecedent_bias=-0.5,
+            consequent_scale=0,
+        ):
+            with pyro.plate("sample", size=1):
+                with pyro.poutine.trace() as trace:
+                    model()
+
+    trace.trace.compute_log_prob
+    nodes = trace.trace.nodes
+
+    values = nodes[cons]["value"]
+    log_probs = nodes[f"__cause____consequent_{cons}"]["fn"].log_factor
+
+    assert log_probs.shape == torch.Size([3, 3, 1, 1, 1, 1])
```

Comment:
Consider asserting something about `indices_of(log_probs)` instead of `log_probs.shape` here. The shape is an implementation detail of `IndexPlatesMessenger` / `MultiWorldCounterfactual` and may change.